source/extensions/access_loggers/grpc/grpc_access_log_impl.h

```diff
@@ -1,40 +1,228 @@
+    void onReceiveMessage(std::unique_ptr<ResponseType>&& message) override {
+      const auto& id = message->id();
+
+      switch (message->status()) {
+      case envoy::service::accesslog::v3::CriticalAccessLogsResponse::ACK:
+        parent_.stats_.critical_logs_ack_received_.inc();
+        parent_.stats_.pending_critical_logs_.dec();
+        parent_.client_->onSuccess(id);
+        break;
+      case envoy::service::accesslog::v3::CriticalAccessLogsResponse::NACK:
+        parent_.stats_.critical_logs_nack_received_.inc();
+        parent_.client_->onError(id);
+        break;
+      default:
+        return;
+      }
+    }
+    void onReceiveTrailingMetadata(Http::ResponseTrailerMapPtr&&) override {}
+    void onRemoteClose(Grpc::Status::GrpcStatus, const std::string&) override {
+      parent_.client_->cleanup();
+    }
+
+    GrpcCriticalAccessLogClient& parent_;
+  };
+
+  class InflightMessageTtlManager {
+  public:
+    InflightMessageTtlManager(Event::Dispatcher& dispatcher,
+                              GrpcCriticalAccessLogClientGrpcClientStats& stats,
+                              Grpc::BufferedAsyncClient<RequestType, ResponseType>& client,
+                              std::chrono::milliseconds message_ack_timeout)
+        : dispatcher_(dispatcher), message_ack_timeout_(message_ack_timeout), stats_(stats),
+          client_(client), timer_(dispatcher_.createTimer([this] { callback(); })) {}
+
+    ~InflightMessageTtlManager() { timer_->disableTimer(); }
+
+    void setDeadline(absl::flat_hash_set<uint32_t>&& ids) {
+      auto expires_at = dispatcher_.timeSource().monotonicTime() + message_ack_timeout_;
+      deadline_.emplace(expires_at, std::move(ids));
+      timer_->enableTimer(message_ack_timeout_);
+    }
+
+  private:
+    void callback() {
+      const auto now = dispatcher_.timeSource().monotonicTime();
+      std::vector<MonotonicTime> expired_timepoints;
+      absl::flat_hash_set<uint32_t> expired_message_ids;
+
+      // Extract timeout message ids.
+      auto it = deadline_.lower_bound(now);
+      while (it != deadline_.end()) {
+        for (auto&& id : it->second) {
+          expired_message_ids.emplace(id);
+        }
+        expired_timepoints.push_back(it->first);
+        ++it;
+      }
+
+      // Clear buffered message ids on the set of waiting timeout.
+      for (auto&& timepoint : expired_timepoints) {
+        deadline_.erase(timepoint);
+      }
+
+      // Restore pending messages to buffer due to timeout.
+      for (auto&& id : expired_message_ids) {
+        const auto& message_buffer = client_.messageBuffer();
+
+        if (message_buffer.find(id) == message_buffer.end()) {
+          continue;
+        }
+
+        auto& message = message_buffer.at(id);
+        if (message.first == Grpc::BufferState::PendingFlush) {
+          client_.onError(id);
+          stats_.critical_logs_message_timeout_.inc();
+        }
+      }
```

Comment:
I imagine this can be simplified and become less heavy something like this: 
```suggestion
      const auto now = dispatcher_.timeSource().monotonicTime();
      auto it = deadline_.lower_bound(now);
      const auto& message_buffer = client_.messageBuffer();
      while (it != deadline_.end()) {
        for (auto&& id : it->second) {
          const auto message = message_buffer.find(id);
          if (message_buffer.find(id) == message_buffer.end()) {
            continue;
          }
          if (message->second.first== Grpc::BufferState::PendingFlush) {
            client_.onError(id);
            stats_.critical_logs_message_timeout_.inc();
          }
        }
        ++it;
      }
      deadline_.erase(deadline_.begin(), it);
```